File: deepeval/openai_agents/extractors.py
from openai.types.responses.response_input_item_param import (
    FunctionCallOutput,
    Message,
)
from openai.types.responses.response_output_message_param import Content
from typing import Union, List
from openai.types.responses import (
    ResponseFunctionToolCallParam,
    ResponseOutputMessageParam,
    ResponseInputContentParam,
    ResponseFunctionToolCall,
    ResponseInputItemParam,
    ResponseOutputRefusal,
    EasyInputMessageParam,
    ResponseOutputMessage,
    ResponseOutputItem,
    ResponseOutputText,
)

from deepeval.tracing.attributes import ToolAttributes, LlmAttributes
from deepeval.prompt.prompt import Prompt
from deepeval.tracing.types import (
    AgentSpan,
    ToolSpan,
    BaseSpan,
    LlmSpan,
)
import json

try:
    from agents import MCPListToolsSpanData
    from agents.tracing.span_data import (
        AgentSpanData,
        FunctionSpanData,
        GenerationSpanData,
        ResponseSpanData,
        SpanData,
        HandoffSpanData,
        CustomSpanData,
        GuardrailSpanData,
    )

    openai_agents_available = True
except ImportError:
    openai_agents_available = False


def _check_openai_agents_available():
    if not openai_agents_available:
        raise ImportError(
            "openai-agents is required for this integration. Install it via your package manager"
        )


def update_span_properties(span: BaseSpan, span_data: SpanData):
    _check_openai_agents_available()
    # LLM Span
    if isinstance(span_data, ResponseSpanData):
        update_span_properties_from_response_span_data(span, span_data)
    elif isinstance(span_data, GenerationSpanData):
        update_span_properties_from_generation_span_data(span, span_data)
    # Tool Span
    elif isinstance(span_data, FunctionSpanData):
        update_span_properties_from_function_span_data(span, span_data)
    elif isinstance(span_data, MCPListToolsSpanData):
        update_span_properties_from_mcp_list_tool_span_data(span, span_data)
    # Agent Span
    elif isinstance(span_data, AgentSpanData):
        update_span_properties_from_agent_span_data(span, span_data)
    # Custom Span
    elif isinstance(span_data, HandoffSpanData):
        update_span_properties_from_handoff_span_data(span, span_data)
    elif isinstance(span_data, CustomSpanData):
        update_span_properties_from_custom_span_data(span, span_data)
    elif isinstance(span_data, GuardrailSpanData):
        update_span_properties_from_guardrail_span_data(span, span_data)


########################################################
### LLM Span ###########################################
########################################################


def update_span_properties_from_response_span_data(
    span: LlmSpan,
    span_data: ResponseSpanData,
):
    response = span_data.response
    if response is None:
        span.model = "NA"
        return
    # Extract prompt template
    prompt_template = response.instructions or None
    prompt = Prompt(template=prompt_template) if prompt_template else None
    # Extract usage tokens
    usage = response.usage
    if usage:
        output_tokens = usage.output_tokens
        input_tokens = usage.input_tokens
        cached_input_tokens = usage.input_tokens_details.cached_tokens
        ouptut_reasoning_tokens = usage.output_tokens_details.reasoning_tokens
    # Get input and output
    input = parse_response_input(span_data.input)
    raw_output = parse_response_output(response.output)
    output = (
        raw_output if isinstance(raw_output, str) else json.dumps(raw_output)
    )
    # Update Span
    llm_attributes = LlmAttributes(
        prompt=prompt,
        input_token_count=input_tokens,
        output_token_count=output_tokens,
        input=input,
        output=output,
    )
    metadata = {
        "cached_input_tokens": cached_input_tokens,
        "ouptut_reasoning_tokens": ouptut_reasoning_tokens,
    }
    span.set_attributes(llm_attributes)
    span.metadata = metadata
    span.model = "NA" if response.model is None else str(response.model)
    span.input = input
    span.output = output
    span.name = "LLM Generation"


def update_span_properties_from_generation_span_data(
    span: LlmSpan,
    generation_span_data: GenerationSpanData,
):
    # Extract usage tokens
    usage = generation_span_data.usage
    if usage:
        output_tokens = usage.get("output_tokens")
        input_tokens = usage.get("input_tokens")
    # Get input and output
    input = generation_span_data.input
    raw_output = generation_span_data.output
    output = (
        raw_output if isinstance(raw_output, str) else json.dumps(raw_output)
    )
    # Update span
    llm_attributes = LlmAttributes(
        prompt=None,
        input_token_count=input_tokens,
        output_token_count=output_tokens,
        input=input,
        output=output,
    )
    span.set_attributes(llm_attributes)
    span.model = generation_span_data.model or "NA"
    span.input = input
    span.output = output
    span.name = "LLM Generation"


########################################################
### Tool Span ##########################################
########################################################


def update_span_properties_from_function_span_data(
    span: ToolSpan,
    function_span_data: FunctionSpanData,
):
    # Update Span
    tool_attributes = ToolAttributes(
        input_parameters=json.loads(function_span_data.input)
        or {"input": function_span_data.input},
        output=function_span_data.output,
    )
    span.set_attributes(tool_attributes)
    span.name = (
        "Function tool: " + function_span_data.name
        if function_span_data.name
        else "Function tool"
    )
    span.description = "Function tool"


def update_span_properties_from_mcp_list_tool_span_data(
    span: ToolSpan,
    mcp_list_tool_span_data: MCPListToolsSpanData,
):
    # Update Span
    tool_attributes = ToolAttributes(
        input_parameters=None, output=mcp_list_tool_span_data.result
    )
    span.set_attributes(tool_attributes)
    span.name = (
        "MCP tool: " + mcp_list_tool_span_data.server
        if mcp_list_tool_span_data.server
        else "MCP tool"
    )
    span.description = "MCP tool"


########################################################
### Agent Span #########################################
########################################################


def update_span_properties_from_agent_span_data(
    span: AgentSpan, agent_span_data: AgentSpanData
):
    # Update Span
    metadata = {}
    span.agent_handoffs = agent_span_data.handoffs
    span.available_tools = agent_span_data.tools
    span.name = agent_span_data.name
    if agent_span_data.output_type:
        metadata["output_type"] = agent_span_data.output_type
    span.metadata = metadata
    span.input = None
    span.output = None


########################################################
### Custom Span #######################################
########################################################


def update_span_properties_from_handoff_span_data(
    span: AgentSpan, handoff_span_data: HandoffSpanData
):
    # Update Span
    metadata = {
        "from_agent": handoff_span_data.from_agent,
        "to_agent": handoff_span_data.to_agent,
    }
    span.name = "Handoff → " + handoff_span_data.to_agent
    span.metadata = metadata
    span.input = None
    span.output = None


def update_span_properties_from_custom_span_data(
    span: BaseSpan, custom_span_data: CustomSpanData
):
    # Update Span
    span.name = custom_span_data.name
    span.metadata = {"data": custom_span_data.data}


def update_span_properties_from_guardrail_span_data(
    span: BaseSpan, guardrail_span_data: GuardrailSpanData
):
    # Update Span
    span.name = "Guardrail: " + guardrail_span_data.name
    span.metadata = {
        "data": guardrail_span_data.triggered,
        "type": guardrail_span_data.type,
    }


########################################################
### Parse Input Utils ##################################
########################################################


def parse_response_input(input: Union[str, List[ResponseInputItemParam]]):
    if isinstance(input, str):
        return input
    processed_input = []
    for item in input:
        if "type" not in item:
            if "role" in item and "content" in item:
                processed_input.append(
                    {
                        "type": "message",
                        "role": item["role"],
                        "content": item["content"],
                    }
                )
        elif item["type"] == "message":
            parsed_message = parse_message_param(item)
            if parsed_message:
                processed_input.append(parsed_message)
        elif item["type"] == "function_call":
            processed_input.append(parse_function_tool_call_param(item))
        elif item["type"] == "function_call_output":
            processed_input.append(parse_function_call_output(item))
    return processed_input if processed_input else None


def parse_message_param(
    message: Union[
        EasyInputMessageParam,
        Message,
        ResponseOutputMessageParam,
    ],
):
    role = message["role"]
    content = message.get("content")
    if isinstance(content, str):
        return {"role": role, "content": content}
    elif isinstance(content, List):
        return {"role": role, "content": parse_message_content_list(content)}
    else:
        return None


def parse_message_content_list(
    content_list: List[Union[ResponseInputContentParam, Content]],
):
    processed_content_list = []
    for item in content_list:
        if item["type"] == "input_text" or item["type"] == "output_text":
            processed_content_list.append(
                {"type": "text", "text": item["text"]}
            )
        elif item["type"] == "input_image":
            # TODO
            ...
        elif item["type"] == "input_file":
            # TODO
            ...
        elif item["type"] == "refusal":
            processed_content_list.append(
                {"type": "refusal", "refusal": item["refusal"]}
            )
    return processed_content_list if processed_content_list else None


def parse_function_tool_call_param(
    tool_call_param: ResponseFunctionToolCallParam,
):
    return {
        "call_id": tool_call_param["call_id"],
        "name": tool_call_param["name"],
        "arguments": tool_call_param["arguments"],
    }


def parse_function_call_output(
    function_call_output: FunctionCallOutput,
):
    return {
        "role": "tool",
        "call_id": function_call_output["call_id"],
        "output": function_call_output["output"],
    }


########################################################
### Parse Output Utils ##################################
########################################################


def parse_response_output(response: List[ResponseOutputItem]):
    processed_output = []
    for item in response:
        if item.type == "message":
            message = parse_message(item)
            if isinstance(message, str):
                processed_output.append(message)
            elif isinstance(message, list):
                processed_output.extend(message)
        elif item.type == "function_call":
            processed_output.append(parse_function_call(item))
    if len(processed_output) == 1:
        return processed_output[0]
    return processed_output if processed_output else None


def parse_message(
    message: ResponseOutputMessage,
) -> Union[str, List[str]]:
    processed_content = []
    for item in message.content:
        if isinstance(item, ResponseOutputText):
            processed_content.append(item.text)
        elif isinstance(item, ResponseOutputRefusal):
            processed_content.append(item.refusal)
    if len(processed_content) == 1:
        return processed_content[0]
    return processed_content if processed_content else None


def parse_function_call(
    function_call: ResponseFunctionToolCall,
):
    return {
        "call_id": function_call.call_id,
        "name": function_call.name,
        "arguments": function_call.arguments,
    }

File: typescript/src/integrations/openai-agents/extractors.ts
import {
  BaseSpan,
  LlmSpan,
  ToolSpan,
  AgentSpan,
  Trace,
} from "../../tracing/tracing";

import {
  SpanData,
  ResponseSpanData,
  GenerationSpanData,
  FunctionSpanData,
  AgentSpanData,
  HandoffSpanData,
  CustomSpanData,
  GuardrailSpanData,
} from "@openai/agents";

function makeJsonSerializable(obj: any): string {
  if (typeof obj === "string") return obj;
  try {
    return JSON.stringify(obj);
  } catch {
    return String(obj);
  }
}

export function updateTracePropertiesFromSpanData(
  trace: Trace,
  spanData: any,
): void {
  if (!trace) return;

  // 1. Give priority to Agent span inputs/outputs for the whole Trace
  if (spanData.type === "agent") {
    if (!trace.input && spanData.input) {
      trace.input = spanData.input;
    }
    if (spanData.output) {
      const rawOutput = spanData.output;
      trace.output =
        typeof rawOutput === "string" ? rawOutput : JSON.stringify(rawOutput);
    }
  }

  // 2. LLM Response
  else if (spanData.type === "response") {
    const data = spanData;
    const response = data._response || data.response;
    const input = data._input || data.input;

    if (!trace.input) {
      trace.input = parseResponseInput(input, response?.instructions);
    }

    if (!trace.output) {
      const rawOutput = parseResponseOutput(response?.output);
      if (rawOutput != null) {
        trace.output =
          typeof rawOutput === "string" ? rawOutput : JSON.stringify(rawOutput);
      }
    }
  }

  // 3. LLM Generation
  else if (spanData.type === "generation") {
    if (!trace.input) {
      trace.input = spanData.input;
    }

    // Add safety check so we don't overwrite Agent output
    if (!trace.output) {
      const rawOutput = spanData.output;
      if (rawOutput != null) {
        trace.output =
          typeof rawOutput === "string" ? rawOutput : JSON.stringify(rawOutput);
      }
    }
  }

  // 4. Function/Tool Execution
  if (spanData.type === "function" || spanData.type === "mcp_tools") {
    if (!trace.toolsCalled) {
      trace.toolsCalled = [];
    }

    let inputParams;
    try {
      inputParams = JSON.parse(spanData.input);
    } catch {
      inputParams = { input: spanData.input };
    }

    trace.toolsCalled.push({
      name: spanData.name || "Function tool",
      inputParameters: inputParams,
      output: spanData.output,
      description: spanData.description,
    });
  }
}

export function updateSpanProperties(span: BaseSpan, spanData: SpanData): void {
  // 1. LLM Spans (Response & Generation)
  if (spanData.type === "response") {
    updateSpanPropertiesFromResponseSpanData(span as LlmSpan, spanData);
  } else if (spanData.type === "generation") {
    updateSpanPropertiesFromGenerationSpanData(span as LlmSpan, spanData);
  }

  // 2. Tool Spans
  else if (spanData.type === "function") {
    updateSpanPropertiesFromFunctionSpanData(span as ToolSpan, spanData);
  } else if (spanData.type === "mcp_tools") {
    updateSpanPropertiesFromMcpListToolsSpanData(span as ToolSpan, spanData);
  }

  // 3. Agent Spans
  else if (spanData.type === "agent") {
    updateSpanPropertiesFromAgentSpanData(span as AgentSpan, spanData);
  }

  // 4. Custom / Handoff / Guardrail Spans
  else if (spanData.type === "handoff") {
    updateSpanPropertiesFromHandoffSpanData(span as AgentSpan, spanData);
  } else if (spanData.type === "custom") {
    updateSpanPropertiesFromCustomSpanData(span, spanData);
  } else if (spanData.type === "guardrail") {
    updateSpanPropertiesFromGuardrailSpanData(span, spanData);
  }
}

// ==========================================================================
// LLM Span Updaters
// ==========================================================================

function updateSpanPropertiesFromResponseSpanData(
  span: LlmSpan,
  spanData: ResponseSpanData,
): void {
  const response = spanData._response;
  if (!response) {
    span.model = "NA";
    return;
  }

  // Extract usage tokens
  const usage = response.usage;
  let cachedInputTokens: number | undefined;
  let outputReasoningTokens: number | undefined;

  if (usage) {
    span.outputTokenCount = usage.output_tokens;
    span.inputTokenCount = usage.input_tokens;

    // Check for detailed usage if available
    if (usage.input_tokens_details) {
      cachedInputTokens = usage.input_tokens_details.cached_tokens;
    }
    if (usage.output_tokens_details) {
      outputReasoningTokens = usage.output_tokens_details.reasoning_tokens;
    }
  }

  // Parse Input and Output
  const input = parseResponseInput(spanData._input, response.instructions);
  const rawOutput = parseResponseOutput(response.output);
  const output =
    rawOutput != null
      ? typeof rawOutput === "string"
        ? rawOutput
        : JSON.stringify(rawOutput)
      : undefined;

  // Update Span Fields
  span.input = input;
  span.output = output;
  span.model = response.model || "NA";
  span.name = "LLM Generation";

  span.metadata = {
    ...(span.metadata || {}),
    cached_input_tokens: cachedInputTokens,
    output_reasoning_tokens: outputReasoningTokens,
  };

  const responseDict = response;
  const invocationParams: Record<string, any> = {};

  const keysToCapture = [
    "max_output_tokens",
    "parallel_tool_calls",
    "reasoning",
    "text",
    "temperature",
    "tool_choice",
    "tools",
    "top_p",
    "truncation",
  ];

  keysToCapture.forEach((key) => {
    if (key in responseDict) {
      invocationParams[key] = responseDict[key];
    }
  });

  span.metadata["invocation_params"] = invocationParams;
}

function updateSpanPropertiesFromGenerationSpanData(
  span: LlmSpan,
  spanData: GenerationSpanData,
): void {
  const usage = spanData.usage;
  if (usage) {
    span.outputTokenCount = usage.output_tokens;
    span.inputTokenCount = usage.input_tokens;
  }

  const input = spanData.input;
  const rawOutput = spanData.output;
  const output =
    rawOutput != null
      ? typeof rawOutput === "string"
        ? rawOutput
        : JSON.stringify(rawOutput)
      : undefined;

  span.model = spanData.model || "NA";
  span.input = input;
  span.output = output;
  span.name = "LLM Generation";

  if (spanData.model_config) {
    span.metadata = {
      ...(span.metadata || {}),
      invocation_params: {
        model_config: makeJsonSerializable(spanData.model_config),
      },
    };
  }
}

// ==========================================================================
// Tool Span Updaters
// ==========================================================================

function updateSpanPropertiesFromFunctionSpanData(
  span: ToolSpan,
  spanData: FunctionSpanData,
): void {
  try {
    span.input = JSON.parse(spanData.input);
  } catch {
    span.input = { input: spanData.input };
  }

  span.output = spanData.output;
  span.name = spanData.name
    ? `Function tool: ${spanData.name}`
    : "Function tool";
  span.description = "Function tool";
}

// ==========================================================================
// Agent Span Updaters
// ==========================================================================

function updateSpanPropertiesFromAgentSpanData(
  span: AgentSpan,
  spanData: AgentSpanData,
): void {
  span.name = spanData.name;
  span.agentHandoffs = spanData.handoffs ? [...spanData.handoffs] : [];
  span.availableTools = spanData.tools ? [...spanData.tools] : [];

  // Populate input/output directly from spanData when available.
  // Cast to any because the SDK type definition omits these runtime fields.
  const rawInput = (spanData as any).input;
  if (rawInput != null) {
    span.input = rawInput;
  }

  const rawOutput = (spanData as any).output;
  if (rawOutput != null) {
    span.output =
      typeof rawOutput === "string" ? rawOutput : JSON.stringify(rawOutput);
  }

  const metadata: Record<string, any> = { ...(span.metadata || {}) };
  if (spanData.output_type) {
    metadata["output_type"] = spanData.output_type;
  }
  span.metadata = metadata;
}

function updateSpanPropertiesFromMcpListToolsSpanData(
  span: ToolSpan,
  spanData: any,
): void {
  span.input = null;
  span.output = spanData.result;
  span.name = spanData.server ? `MCP tool: ${spanData.server}` : "MCP tool";
  span.description = "MCP tool";
}

// ==========================================================================
// Custom / Handoff / Guardrail Updaters
// ==========================================================================

function updateSpanPropertiesFromHandoffSpanData(
  span: AgentSpan,
  spanData: HandoffSpanData,
): void {
  span.name = `Handoff → ${spanData.to_agent}`;
  span.input = null; // Handoffs usually don't have distinct IO in this context
  span.output = null;
  span.metadata = {
    ...(span.metadata || {}),
    from_agent: spanData.from_agent,
    to_agent: spanData.to_agent,
  };
}

function updateSpanPropertiesFromCustomSpanData(
  span: BaseSpan,
  spanData: CustomSpanData,
): void {
  span.name = spanData.name;
  span.metadata = {
    ...(span.metadata || {}),
    data: spanData.data,
  };
}

function updateSpanPropertiesFromGuardrailSpanData(
  span: BaseSpan,
  spanData: GuardrailSpanData,
): void {
  span.name = `Guardrail: ${spanData.name}`;
  span.metadata = {
    ...(span.metadata || {}),
    triggered: spanData.triggered,
    type: spanData.type,
  };
}

// ==========================================================================
// Input Parsing Utilities
// ==========================================================================

export function parseResponseInput(
  input: any,
  instructions: any,
): any[] | null {
  const processedInput = [];

  // Case 1: Both are strings -> Create System + User messages
  if (typeof input === "string" && typeof instructions === "string") {
    return [
      { type: "message", role: "system", content: instructions },
      { type: "message", role: "user", content: input },
    ];
  }
  // Case 2: Both are lists -> Concatenate
  else if (Array.isArray(input) && Array.isArray(instructions)) {
    // Treat instructions as prepended context
    const combined = [...instructions, ...input];
    return processInputList(combined);
  }
  // Case 3: List input, String instructions
  else if (Array.isArray(input) && typeof instructions === "string") {
    processedInput.push({
      type: "message",
      role: "system",
      content: instructions,
    });
    // Process the rest of the input list
    const remaining = processInputList(input);
    if (remaining) processedInput.push(...remaining);
    return processedInput;
  }
  // Case 4: String input, List instructions
  else if (typeof input === "string" && Array.isArray(instructions)) {
    const processedInstructions = processInputList(instructions);
    if (processedInstructions) processedInput.push(...processedInstructions);
    processedInput.push({ type: "message", role: "user", content: input });
    return processedInput;
  }

  // Fallback: simple processing of input if it matches none above exactly
  if (Array.isArray(input)) {
    return processInputList(input);
  }

  return null;
}

function processInputList(itemList: any[]): any[] | null {
  const processed: any[] = [];

  for (const item of itemList) {
    // If it's a raw dict with role/content but no type, standardise it
    if (!item.type && item.role && item.content) {
      processed.push({
        type: "message",
        role: item.role,
        content: item.content,
      });
      continue;
    }

    if (item.type === "message") {
      const parsed = parseMessageParam(item);
      if (parsed) processed.push(parsed);
    } else if (item.type === "function_call") {
      processed.push(parseFunctionToolCallParam(item));
    } else if (item.type === "function_call_output") {
      processed.push(parseFunctionCallOutput(item));
    }
  }

  return processed.length > 0 ? processed : null;
}

function parseMessageParam(message: any): any {
  const role = message.role;
  const content = message.content;

  if (typeof content === "string") {
    return { type: "message", role, content };
  } else if (Array.isArray(content)) {
    return { type: "message", role, content: parseMessageContentList(content) };
  }
  return null;
}

function parseMessageContentList(contentList: any[]): any[] | null {
  const processed: any[] = [];
  for (const item of contentList) {
    if (
      item.type === "input_text" ||
      item.type === "output_text" ||
      item.type === "text"
    ) {
      processed.push({ type: "text", text: item.text });
    } else if (item.type === "refusal") {
      processed.push({ type: "refusal", refusal: item.refusal });
    }
  }
  return processed.length > 0 ? processed : null;
}

function parseFunctionToolCallParam(toolCall: any): any {
  return {
    call_id: toolCall.call_id,
    name: toolCall.name,
    arguments: toolCall.arguments,
  };
}

function parseFunctionCallOutput(output: any): any {
  return {
    role: "tool",
    call_id: output.call_id,
    output: output.output,
  };
}
// ==========================================================================
// Output Parsing Utilities
// ==========================================================================

export function parseResponseOutput(responseItems: any): any {
  if (typeof responseItems === "string") return responseItems;
  if (!responseItems) return null;

  if (!Array.isArray(responseItems)) {
    if (typeof responseItems.content === "string") return responseItems.content;
    if (typeof responseItems.output === "string") return responseItems.output;
    return null;
  }

  const processedOutput: any[] = [];
  for (const item of responseItems) {
    if (typeof item === "string") {
      processedOutput.push(item);
    } else if (
      item.type === "message" ||
      item.role === "assistant" ||
      item.status === "completed"
    ) {
      const message = parseMessageOutput(item);
      if (message) {
        if (typeof message === "string") {
          processedOutput.push(message);
        } else if (Array.isArray(message)) {
          processedOutput.push(...message);
        }
      }
    } else if (item.type === "function_call") {
      processedOutput.push(parseFunctionCall(item));
    }
  }

  if (processedOutput.length === 1) return processedOutput[0];
  return processedOutput.length > 0 ? processedOutput : null;
}

function parseMessageOutput(message: any): string | string[] | null {
  if (!message.content) return null;

  // If the SDK returned a flat string, return it immediately
  if (typeof message.content === "string") {
    return message.content;
  }

  // If it's an array, map through the text blocks
  if (Array.isArray(message.content)) {
    const processedContent: string[] = [];

    for (const item of message.content) {
      if (typeof item === "string") {
        processedContent.push(item);
      } else if (
        (item.type === "text" || item.type === "output_text") &&
        item.text
      ) {
        processedContent.push(item.text);
      } else if (item.type === "refusal" && item.refusal) {
        processedContent.push(item.refusal);
      }
    }

    if (processedContent.length === 1) return processedContent[0];
    return processedContent.length > 0 ? processedContent : null;
  }

  return null;
}

function parseFunctionCall(functionCall: any): any {
  return {
    call_id: functionCall.call_id,
    name: functionCall.name,
    arguments: functionCall.arguments,
  };
}
